feat(dart/transform): Introduce @angularEntrypoint

kegluneq opened the pull request:
@yjbanov, this is the annotation declaration as I imagine it -- any objections? One result of this being Dart-only is that the examples in `playground/` cannot use it, since the annotation is not available to the .ts code.

Is the idea behind this annotation that the user will no longer need to specify `entry_points` to the transformer?

---

Add an annotation which the transformer uses to determine where to:
- Insert calls to `initReflector`
- Rewrite `import '.../bootstrap.dart'` => `bootstrap_static.dart`
- Rewrite `bootstrap(Type)` => `bootstrapStatic(Type)`

Closes #4865

yjbanov commented:
I think we should make this annotation available to TS code that we transpile to Dart. Angular 2 JS users wouldn't use it, but we would.

kegluneq commented:
I'm fine with that as long as Misko is - not sure what he'll think in regards to his "this should be dart-specific" comment

kegluneq commented:
/cc @mhevery any objections to baking the annotation into TS code? In your comment on #4865 you seemed to oppose it

Commit:
feat(dart/transform): Introduce @AngularEntrypoint()

yjbanov commented on modules/angular2/src/platform/browser_common.ts:
```diff
@@ -77,4 +77,24 @@ export function initDomAdapter() {
+ * void main() {
+ *   bootstrap(MyComponent);
+ * }
+ */
```
How about we move the docs to the constant as that's where IDE code navigation will takes users.

yjbanov commented on modules/angular2/src/platform/browser_common.ts:
```diff
@@ -77,4 +77,24 @@ export function initDomAdapter() {
+export class _AngularEntrypoint {
+  constructor() {}
+}
+const angularEntrypoint = CONST_EXPR(new _AngularEntrypoint());
```
Let's move the constant to the top of the file to make it more prominent.

yjbanov commented on modules/angular2/src/platform/browser_common.ts:
```diff
@@ -77,4 +77,24 @@ export function initDomAdapter() {
+ */
+@CONST()
+export class _AngularEntrypoint {
+  constructor() {}
```
I think we could take an optional string parameter that describes the entry point (and therefore make the class public). We could use it in the transformer for logging, e.g.:

```dart
@AngularEntrypoint("application shell")

```

Then:

```
[INFO] transforming entrypoint 'application shell'
```

Commit:
feature(dart/transform): Inject `initReflector` at @AngularEntrypoint

Detect the `@AngularEntrypoint` annotations on methods and/or
functions and add a call to `initReflector` there.

See #4865

kegluneq commented:
I believe this is now ready for review.

kegluneq commented:
I had to remove the last commit (demonstrating `@AngularEntrypoint` in hello_world) because TS doesn't allow decorators on top level functions.

Let's see if Travis is happy now :)

yjbanov commented:
Maybe add it [here](https://github.com/angular/angular/blob/master/modules_dart/payload/hello_world/web/index.dart) instead?

kegluneq commented:
/cc @vsavkin, could you PTAL?

I added lines to public_api_spec.ts as we discussed, but Travis reports an error re: DomAdapter:dart, and I cannot figure out how these changes could have hidden this symbol.

kegluneq commented:
Bug in test fixed at cf157b99d3186f568a2afea10c6cd74f9c81bb44

This should now be ready for merge.

jteplitz602 commented:
@kegluneq Can we move this class outside of `browser_common.ts`? We can't load browser_common on non-browser platforms like WebWorkers or native. @yjbanov and I discussed putting it in `src/core/angular_entrypoint.ts` instead. What do you think?

kegluneq commented:
SGTM, will update this PR.

kegluneq commented:
@yjbanov or @jteplitz602, moved the annotation to `src/core/angular_entrypoint.ts`. PTAL

jteplitz602 commented:
LGTM

kegluneq commented:
Now ready for merge for realz :)

vsavkin commented:
Merged https://github.com/angular/angular/commit/87449ab43c8547ea786713f5a2245b1ab2197afd
